pkg/scheduler/factory/factory.go

```diff
@@ -970,15 +963,13 @@ func (c *configFactory) deleteNodeFromCache(obj interface{}) {
 		glog.Errorf("cannot convert to *v1.Node: %v", t)
 		return
 	}
-	// NOTE: Because the scheduler uses snapshots of schedulerCache and the live
-	// version of equivalencePodCache, updates must be written to schedulerCache
-	// before invalidating equivalencePodCache.
-	if err := c.schedulerCache.RemoveNode(node); err != nil {
-		glog.Errorf("scheduler cache RemoveNode failed: %v", err)
-	}
+	// NOTE: Invalidate equivalence node cache first.
 	if c.enableEquivalenceClassCache {
 		c.equivalencePodCache.InvalidateAllPredicatesOnNode(node.GetName())
 	}
+	if err := c.schedulerCache.RemoveNode(node); err != nil {
```

Comment:
If remove must happen after the invalidation, it deserves a better comment.

Reply:
Sorry, I found I'm wrong with this part. All equivalence cache invalidations should be after resource objects are added, updated and deleted, then predicate functions will run and use new versions of resource objects. I revert this change.

By the way, there is no need to invalidate predicates for node after adding because no predicate functions will run on deleted node (we iterate on nodes in scheduler cache).

Reply:
I agree with you that invalidation of eCache should happen after CUD operations on objects. The following diagram shows a scenario where the reverse order causes an issue. If invalidation happens first, we could snapshot eCache after the invalidation and then we snapshot the cache itself. A node is then removed/added/updated before we update eCache. We then would update eCache with stale information which is based on our latest snapshot of the cache taken before the node removal/addition/update:

```
------|-----------------|------------------|----------------|---------------|-------->
Invalidate eCache  Snapshot eCache   Snapshot Cache     Remove Node    Update eCache    
```

I think you should add this to the comment to document why the order is important.

Reply:
Added in https://github.com/kubernetes/kubernetes/pull/67308/commits/dd9d576e64f4536ffe8d9367d1d2edc90ee3d3ab.

There is no need to revert PR https://github.com/kubernetes/kubernetes/pull/63895/files, I'll rebase it when PR is done.